module/zfs/zfs_vnops.c

```diff
@@ -1848,7 +1848,7 @@ zfs_remove(struct inode *dip, char *name, cred_t *cr, int flags)
 		zfs_unlinked_add(zp, tx);
 		mutex_exit(&zp->z_lock);
 		zfs_inode_update(zp);
-		iput(ip);
+		zfs_iput_async(ip);
```

Comment:
I think this makes delete_now not really delete now, maybe we can just remove this path.